storage/src/journal/variable.rs

```diff
@@ -1,2223 +1,1443 @@
+        if self.size.is_multiple_of(self.items_per_section) {
+            futures::try_join!(self.data.sync(section), self.offsets.sync())?;
         }
 
-        // If compression is enabled, decompress the item
-        let item = if compressed {
-            let decompressed =
-                decode_all(Cursor::new(&item)).map_err(|_| Error::DecompressionFailed)?;
-            V::decode_cfg(decompressed.as_ref(), cfg).map_err(Error::Codec)?
-        } else {
-            V::decode_cfg(item, cfg).map_err(Error::Codec)?
-        };
+        Ok(position)
+    }
 
-        // Calculate next offset
-        let current_pos = reader.position();
-        let aligned_offset = compute_next_offset(current_pos)?;
-        Ok((aligned_offset, current_pos, size as u32, item))
+    /// Return the total number of items that have been appended to the journal.
+    ///
+    /// This count is NOT affected by pruning. The next appended item will receive this
+    /// position as its value.
+    pub async fn size(&self) -> Result<u64, Error> {
+        Ok(self.size)
     }
 
-    /// Reads an item from the blob at the given offset and of a given size.
-    async fn read_exact(
-        compressed: bool,
-        cfg: &V::Cfg,
-        blob: &Append<E::Blob>,
-        offset: u32,
-        len: u32,
-    ) -> Result<V, Error> {
-        // Read buffer
-        let offset = offset as u64 * ITEM_ALIGNMENT;
-        let entry_size = 4 + len as usize + 4;
-        let buf = blob.read_at(vec![0u8; entry_size], offset).await?;
-
-        // Check size
-        let mut hasher = crc32fast::Hasher::new();
-        let disk_size = u32::from_be_bytes(buf.as_ref()[..4].try_into().unwrap());
-        hasher.update(&buf.as_ref()[..4]);
-        if disk_size != len {
-            return Err(Error::UnexpectedSize(disk_size, len));
+    /// Return the position of the oldest item still retained in the journal.
+    ///
+    /// Returns `None` if the journal is empty or if all items have been pruned.
+    pub async fn oldest_retained_pos(&self) -> Result<Option<u64>, Error> {
+        if self.size == self.oldest_retained_pos {
+            // No items retained: either never had data or fully pruned
+            Ok(None)
+        } else {
+            Ok(Some(self.oldest_retained_pos))
         }
+    }
 
-        // Verify integrity
-        let item = &buf.as_ref()[4..4 + len as usize];
-        hasher.update(item);
-        let checksum = hasher.finalize();
-        let stored_checksum =
-            u32::from_be_bytes(buf.as_ref()[4 + len as usize..].try_into().unwrap());
-        if checksum != stored_checksum {
-            return Err(Error::ChecksumMismatch(stored_checksum, checksum));
+    /// Prune items at positions strictly less than `min_position`.
+    ///
+    /// Returns `true` if any data was pruned, `false` otherwise.
+    ///
+    /// This prunes both the data journal and the offsets journal to maintain consistency.
```

Comment:
another implementation "leak" -- should the user even know about "locations" outside of specifying the location partition in Config?

Reply:
Removed this line.